text/0126-Configurable-Metric-Aggregations.md

```diff
@@ -0,0 +1,95 @@
+# A Proposal For SDK Support for Configurable Batching and Aggregations
+
+Add SDK support for the ability to configure Metric Aggregations.
+
+## Motivation
+
+OpenTelemetry's architecture separates the concerns of instrumentation and operation. The Metric Instruments
+provided by the Metric API are all defined to have a default aggregation. And, by default, aggregations are
+performed with all Labels being used to define a unit of aggregation. Although this is a good default
+configuration for the SDK to provide, more configurability is needed.
+
+There are 3 main use-cases that this proposal is intended to address:
+
+1) The application developer/operator wishes to use an aggregation other than the default provided by the SDK
+for a given instrument or set of instruments.
+2) An exporter author wishes to inform the SDK what "Temporality" (delta vs. cumulative) the resulting metric
+data points represent. "Delta" means only metric recordings since the last reporting interval are considered
+in the aggregation, and "Cumulative" means that all metric recordings over the lifetime of the Instrument are
+considered in the aggregation.
+3) The application developer/operator wishes to constrain the cardinality of labels for metrics being reported
+to the metric vendor/backend of choice.
+
+## Explanation
+
+I propose a new SDK feature, available on the interface of the SDK's MeterProvider implementation, to configure
+the batching strategies and aggregations that will be used by the SDK when metric recordings are made.
+
+The basic API has two parts.
+
+* InstrumentSelector - Enables specifying the selection of one or more instruments for the configuration to apply to.
+  - Selection options include: the instrument type (Counter, ValueRecorder, etc), and a regex for instrument name.
```

Comment:
I get why users need to be able to set things like whether to use a delta or cumulative for all `Counter` instruments and the default aggregator for all `ValueRecorder` instruments, but the selector pattern with regexes feels overly powerful. I would have expected this to be simply SDK configuration like `default_value_recorder_aggregator: ddsketch`.

Reply:
I think there are solid use-cases for only using an expensive aggregation for a subset of named instruments. eg. I'm running an http service, and I want details on only the ValueRecorders with a given naming pattern like "http.*.duration". Whereas, you might want just a very simple summary aggregation for any other timed operations in your application, or something that isn't a timed operation at all, like the size of incoming http request bodies.

Reply:
Agreed about only sometimes wanting certain expensive aggregations, whether it fits a pattern or not is what I'm unsure about. Another example would be wanting specific distribution buckets. I would just have expected them to be set on a specific aggregation of an instrument and not through a regex.

Reply:
well, you can give the specific instrument name as the regex. that would work fine in this scheme. :)

Reply:
My rationale is that the operator won't necessarily know all the details of all the instruments that are baked into the instrumentation an app is using. But, they should be able to use general semantic conventions to pick out what they want. I don't feel super strongly about this, though. What is your concern with this? That it's a footgun, or that regex is horrible and should die in a fire? Or something else?

Reply:
The footgun :). But maybe it isn't so dangerous.

Reply:
Will bring up in the meeting now but putting here for record.

In the case of views each view has a name. In this selection process is this a) creating an aggregation for each metric matched individually or aggregating them together (I assumed the former) b) you can't give an individual name to each "view" selection is done this way.

Reply:
@tsloughter I like this:

> `default_value_recorder_aggregator: ddsketch`

In the experimental branch I hacked on (briefly), [here](https://github.com/jmacd/opentelemetry-go/blob/jmacd/configurable/sdk/metric/integrator/configurable/configurable.go#L42) is the Config struct that has a Defaults sections w/ aggregation by instrument kind, a Views w/ aggregation by instrument name, followed by an Aggregations section that contains Labels/Aggregator configs.